File: fridge-service/src/main/java/com/zenika/tutoriels/observabilitydemo/fridge/FridgeApplication.java
package com.zenika.tutoriels.observabilitydemo.fridge;

import com.zenika.tutoriels.observabilitydemo.fridge.business.CheeseInitialStock;
import com.zenika.tutoriels.observabilitydemo.fridge.business.Fridge;
import com.zenika.tutoriels.observabilitydemo.fridge.business.MeatInitialStock;
import com.zenika.tutoriels.observabilitydemo.fridge.web.CheeseController;
import com.zenika.tutoriels.observabilitydemo.fridge.web.FridgeErrorControllerAdvice;
import com.zenika.tutoriels.observabilitydemo.fridge.web.MeatController;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.SpringBootConfiguration;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.ConstructorBinding;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Profile;

import static java.lang.Math.random;

@SpringBootConfiguration
@EnableAutoConfiguration
@EnableConfigurationProperties({FridgeApplication.FridgeMeatConfigurationProperties.class,
        FridgeApplication.FridgeCheeseConfigurationProperties.class})
public class FridgeApplication {

    public static void main(String[] args) {
        SpringApplication.run(FridgeApplication.class, args);
    }

    @ConfigurationProperties("fridge.meat")
    @ConstructorBinding
    public record FridgeMeatConfigurationProperties(int chicken, int beef, int pork) {
    }

    @ConfigurationProperties("fridge.cheese")
    @ConstructorBinding
    public record FridgeCheeseConfigurationProperties(int cheddar, int goat, int blue) {
    }

    @Bean
    @Profile("!generate-random-failures")
    public Fridge getSafeFridge(FridgeMeatConfigurationProperties meats, FridgeCheeseConfigurationProperties cheeses) {
        return new Fridge(new MeatInitialStock(meats.chicken(), meats.beef(), meats.pork()),
                new CheeseInitialStock(cheeses.cheddar(), cheeses.goat(), cheeses.blue()), () -> {});
    }

    @Bean
    @Profile("generate-random-failures")
    public Fridge getFailingFridge(FridgeMeatConfigurationProperties meats, FridgeCheeseConfigurationProperties cheeses) {
        return new Fridge(new MeatInitialStock(meats.chicken(), meats.beef(), meats.pork()),
                new CheeseInitialStock(cheeses.cheddar(), cheeses.goat(), cheeses.blue()), () -> {
            if(random() > 0.8d){
                throw new RuntimeException("Random error");
            }
        });
    }

    @Bean
    public MeatController meatController(Fridge fridge) {
        return new MeatController(fridge);
    }

    @Bean
    public CheeseController cheeseController(Fridge fridge) {
        return new CheeseController(fridge);
    }

    @Bean
    public FridgeErrorControllerAdvice errorControllerAdvice(){
        return new FridgeErrorControllerAdvice();
    }

}

File: fridge-service/src/main/kotlin/com/zenika/tutoriels/observabilitydemo/fridge/FridgeApplication.kt
package com.zenika.tutoriels.observabilitydemo.fridge

import com.zenika.tutoriels.observabilitydemo.fridge.business.CheeseInitialStock
import com.zenika.tutoriels.observabilitydemo.fridge.business.Fridge
import com.zenika.tutoriels.observabilitydemo.fridge.business.MeatInitialStock
import com.zenika.tutoriels.observabilitydemo.fridge.web.FridgeWebHandler
import org.springframework.boot.SpringApplication
import org.springframework.boot.SpringBootConfiguration
import org.springframework.boot.autoconfigure.EnableAutoConfiguration
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.ConstructorBinding
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.http.MediaType
import org.springframework.web.reactive.function.server.*


@ConfigurationProperties("fridge.meat")
@ConstructorBinding
data class FridgeMeatConfigurationProperties(val chicken: Int, val beef: Int, val pork: Int)

@ConfigurationProperties("fridge.cheese")
@ConstructorBinding
class FridgeCheeseConfigurationProperties(val goat: Int, val cheddar: Int, val blue: Int)


@SpringBootConfiguration
@EnableAutoConfiguration
@EnableConfigurationProperties(FridgeMeatConfigurationProperties::class, FridgeCheeseConfigurationProperties::class)
class FridgeApplication {



    @Bean
    fun fridge(meats: FridgeMeatConfigurationProperties, cheeses: FridgeCheeseConfigurationProperties): Fridge {
        return Fridge(MeatInitialStock(meats.chicken, meats.beef, meats.pork),
                CheeseInitialStock(cheeses.cheddar, cheeses.goat, cheeses.blue))
    }

    @Bean
    fun router(fridge: Fridge): RouterFunction<ServerResponse> {
        val handler = FridgeWebHandler(fridge)
        return router {
            accept(MediaType.APPLICATION_JSON)
                    .nest {
                        GET("/meats").invoke(handler::listMeats)
                        DELETE("/meats/{kind}").invoke(handler::orderMeatByKind)
                        GET("/cheeses").invoke(handler::listCheeses)
                        DELETE("/cheeses/{kind}").invoke(handler::orderCheeseByKind)
                    }
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            SpringApplication.run(FridgeApplication::class.java, *args)
        }
    }
}